AOC [20.02](https://adventofcode.com/2020/day/2)

Puzzle 1: corrupted password database

Given list of inputs:
pt1: policy (min-max number of character)
pt2: password

For each line, is the password meet the criteria of the policy?

How many PWs meet the criteria?

Solution:
1. Read each line and split on ":" char
1. Read the head and look for min/max/char
1. read the pw and count the number of instances of the char: is it between min and max?

In [10]:
line = '1-3 a: abaaade'

head,_,tail = line.partition(':')
head,_,testChar = head.partition(' ')
minVal,_,maxVal = head.partition('-')

print(minVal,maxVal,testChar)
countChar = int(tail.count(testChar))
if (int(minVal) < countChar < int(maxVal)):
    print('PW valid')
else:
    print('PW invalid')


1 3 a
PW invalid


In [13]:
testData = '''1-3 a: abcde
1-3 b: cdefg
2-9 c: ccccccccc'''

validPWs = 0

for line in testData.split('\n'):
    head,_,tail = line.partition(':')
    head,_,testChar = head.partition(' ')
    minVal,_,maxVal = head.partition('-')
    countChar = int(tail.count(testChar))
    
    print(minVal,maxVal,testChar,countChar)
    if (int(minVal) <= countChar <= int(maxVal)):
        validPWs += 1

print(f"{validPWs} valid PWs found")

1 3 a 1
1 3 b 0
2 9 c 9
2 valid PWs found


In [14]:
with open ('2002input.txt') as f_in:
    data = [line for line in f_in.read().split('\n')]

validPWs = 0

for line in data:
    head,_,tail = line.partition(':')
    head,_,testChar = head.partition(' ')
    minVal,_,maxVal = head.partition('-')

    countChar = int(tail.count(testChar))
    if (int(minVal) <= countChar <= int(maxVal)):
        validPWs += 1

print(f"{validPWs} valid PWs found")


569 valid PWs found


569 is correct.

Puzzle 2:
The two digits don't indicate min/max values, but index positions (indexing beg at 1 not zero!). PW is valid if the testChar appears in the string at either of the positions (but not both).

Solution:
1. keep all the code, but change the test!

In [20]:
testData = '''1-3 a: abcde
1-3 b: cdefg
2-9 c: ccccccccc'''

validPWs = 0

for line in testData.split('\n'):
    head,_,tail = line.partition(':')
    head,_,testChar = head.partition(' ')
    minVal,_,maxVal = head.partition('-')
    tail = tail.strip()
    
    print(tail.strip(),tail[int(minVal)-1],tail[int(maxVal)-1],testChar)

    if (tail[int(minVal)-1] == testChar) or (tail[int(maxVal)-1] == testChar):
        validPWs += 1

print(f"{validPWs} valid PWs found")

abcde a c a
cdefg c e b
ccccccccc c c c
2 valid PWs found


In [21]:
with open ('2002input.txt') as f_in:
    data = [line for line in f_in.read().split('\n')]

validPWs = 0

for line in data:
    head,_,tail = line.partition(':')
    head,_,testChar = head.partition(' ')
    minVal,_,maxVal = head.partition('-')
    tail = tail.strip()
    
    #print(tail.strip(),tail[int(minVal)-1],tail[int(maxVal)-1],testChar)

    if (tail[int(minVal)-1] == testChar) or (tail[int(maxVal)-1] == testChar):
        validPWs += 1

print(f"{validPWs} valid PWs found")

671 valid PWs found


Too many. This is because the OR is finding some PWs where BOTH indexes have the test character. 

In [22]:
with open ('2002input.txt') as f_in:
    data = [line for line in f_in.read().split('\n')]

validPWs = 0

for line in data:
    head,_,tail = line.partition(':')
    head,_,testChar = head.partition(' ')
    minVal,_,maxVal = head.partition('-')
    tail = tail.strip()
    
    testString = tail[int(minVal)-1] + tail[int(maxVal)-1] # pull out the two index characters
    if int(testString.count(testChar)) == 1: # count how many match the test character
        validPWs += 1

print(f"{validPWs} valid PWs found")

346 valid PWs found


That's correct.